iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp

```diff
@@ -673,9 +675,15 @@ void ConvertToLLVMPass::runOnOperation() {
   });
 
   OwningRewritePatternList patterns(&getContext());
+  // TOSA rescale will be lowered to Standard right before LLVM codegen. Future
+  // lowerings should using specific CPU operations to avoid requiring i64
+  // values.
+  tosa::populateTosaRescaleToStandardConversionPatterns(&patterns);
```

Comment:
Do we have a conversion pass that runs before both "LinalgToLLVM" and "LinalgToSPIRV"? Neither of these files looks like the right place for a "TosaToStandard" conversion to me. This would also run during our conversion from different frontend dialects (e.g. HLO)... wouldn't we rather push frontend conversions up a level?

Reply:
So this is specifically for a TOSA operation that occurs within Linalg bodies. We need to perform this at the per-backend in level as future versions should support specific CPU intrinsics. Initially I performed this part of the lowering during the TosaToLinalg conversion, however that wouldn't handle separate lowerings per-backend.

Reply:
Can you post some IR snippets around this conversion flow?

Reply:
Added more thorough comment with explanation why it works.

Reply:
That's not quite what I had in mind...

I want to discuss _why these patterns exist here_, on this PR, with context for why TOSA ops appear inside linalg bodies. Comments in the code are nice, but that comment is too specific (and would probably need to be mirrored between both codegen files).

For example, I'd like to see the `-print-ir-after-all` before/after adding these patterns during conversion, possibly edited to just include the relevant ops. It's still not clear to me what is happening with these dialects. I'm less concerned with the specific implementation of a particular bundle of arithmetic and more with how the IR is being modified.

Reply:
The most concise explanation I can provide is the ApplyScale, unlike other TOSA Ops, is a utility operation for supporting lowering. TOSA multiply, round, shift behavior is not expressible in StandardOps without bit-extending to i64 types.

Specifically this Op is constructed during the TOSA to Linalg lowerings and is placed without linalg.generic bodies. It is not expanded to StandardOps at that point so that different targets can support custom intrinsics that are not expressible our StandardOps/LLVM dialects.

As for specifying the ir-after-all, the difference is that a single tosa.apply_scale(...) op is expanded to the pseudo-IR shown.

Reply:
I'm still not _quite_ following, but this is helping. Thanks for your patience. So `LinalgToLLVM/ConvertToLLVM.cpp` and `LinalgToSPIRV/ConvertToSPIRVPass.cpp` will diverge? I think the comments should speak to that more directly.

I don't think the "This can be performed using i64 types as" comment in `LinalgToLLVM/ConvertToLLVM.cpp` is useful as-is. It's too specific when surrounded by other lowerings that are just "populateVectorToSCFConversionPatterns" (which itself probably does stuff at least as complicated as this :P) and could be replaced with something like
```
// Use the default lowering for TOSA's ApplyScale operator.
// <_brief_ 1-2 sentence description of how bit extending factors into this?>
// TODO(name): Add lowering that uses custom intrinsics specific to LLVM/SPIRV.
```

Reply:
This looks weird to me too. TOSA dialect should be higher level IR to me. If something is missing in Standard/Math/SomeQuant dialect, we should add such support. Leaving a TOSA op until the latest stage looks weird to me, and I'm afraid if it will cause more maintenance.

Reply:
@ScottTodd updated following your preferences mapping to silvas and antiagainst as each advised on their respective specializations.

@hanhanW Silvas and Antiagainst specifically advised in favor of lowering to target specific behavior for ApplyScale, noting that long term a quantization specific dialect should exists. However with only a single use-case, we cannot demonstrate a sufficient need for a new dialect in MLIR. We expect this to change in the future, however it is not expected to happen anytime soon.